### DISCUSSION TITLE
Etymology of the term Folding

### DISCUSSION THREAD

**Discussion by stephan-j-lange** [rust-unofficial/patterns#383] [General]
First of all thanks a lot for this great documentation. It helps our team a lot.

In ***Design Patterns / Creational / Fold*** you write "The etymology here is unclear to me".

I believe that the term comes from digital signal processing. See https://en.wikipedia.org/wiki/Folding_(DSP_implementation).

When I was working in the area of image processing then we used the term of folding for example when talking about operations that shall be executed for each pixel in relation to the neighboring pixels. Examples are edge detection or noise reduction.

When reading your nice description I see the similarities.

**Comment by neithernut** [comment]
I suspect it started out as a synonym for or specific kind of [convolution](https://en.wikipedia.org/wiki/Convolution), as in integral over a function product. At least in German, you'd call this type of convolution "Faltung", which you can translate as "fold".

Also note that, while the Wikipedia article https://en.wikipedia.org/wiki/Fold_(higher-order_function) mainly talks about functional programming languages, the term was likely already used for some lambda calculus in the 30s, which predates at least _digital_ signal processing.
